[MODULES-3887] Implement iis_site flush method

wilson208 opened the pull request:
Within this PR is the implementation of the flush method for the iis_site resource and accompanying acceptance tests to cover this. A number of small fixes were made as required.

**Implement flush for iis_site resource’s web administration provider:**
By reusing a lot of the same powershell code used when creating a new iis_site resource I implemented the `update` method within the iis_site webadministration provider. I abstracted the powershell Try-SetItemProperty to a new template file to avoid repetition.

**Acceptance Tests**
The following tests have been added for the iis_site resource and utility methods implemented to support these tests:
- Create a site with basic params
- Create a site with all params specified
  - Using W3C log format, logflags and logtruncatesize
  - Using NCSA log format and logperiod rather than logtruncatesize
  Change site state
  - From Stopped to Started
  - From Started to Stopped
  - From Started to Absent
- Ensure run fails with invalid params for:
  - logformat param
  - logperiod param
  Ensure can update existing, previously set param:
  - physicalpath param
  - enabledprotocols param
  - logflags param
- Utility methods added:
  - remove_all_sites
  - create_site(name, started, path)
  - create_path(path)

**Fixes:**
- iis_site webadministration provider:
  - Fix errors when no iis_site resources exist
  - Fix typo causing self.instances to return empty physicalpath's
  - In iis_site’s self.to_bool, handle case of empty value returns false in existing regex.
  - In iis_sites’s self.to_bool, return true or false symbols
  - Fix logproperties.ps1 template to correctly set array of flags as a comma-separated string when iis_site is being created or updated

- iis_site type:
  - Overwrite logflags insync? method to deal with differences in ordering of array

- Acceptance Test Support Files:
  - Standardise how we test for idempotency in the `idempotent_resource` shared example, so it is consistent with other modules
  - Escape values passed to `puppet_resource_should_show` to allow validation of path strings containing special characters (e.g. backslashes)

jpogran commented on lib/puppet/provider/iis_site/iisadministration.rb:
```diff
@@ -1,23 +0,0 @@
-require 'pathname'
-
-Puppet::Type.type(:iis_site).provide(:iisadministration) do
```
This file shouldn't be deleted

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -9,8 +9,8 @@
 
-  confine    :iis_version     => ['8.0','8.5']
-  confine    :operatingsystem => [:windows ]
+  confine :iis_version => ['8.0', '8.5']
```
Is there a reason for the whitespace changes here?

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -38,6 +38,59 @@ def create
   end
 
+  def update
+    Puppet.err 'UPDATE FUNCTION'
```
TODO: remove

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -38,6 +38,59 @@ def create
+    inst_cmd << '$serverManager.CommitChanges();'
+    inst_cmd = inst_cmd.join
+
+    Puppet.err('COMMAND: ' + inst_cmd)
```
Puppet.debug works nicer here

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -38,6 +38,59 @@ def create
+    result = self.class.run(inst_cmd)
+    Puppet.err('RESULT: ' + result[:stdout])
+
+    # Handle errors - had issue where when an issue occurred, errors were only output to stdout, not stderr
```
This is powershell, not bash. We have to handle errors. See comment about erb above

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -38,6 +38,59 @@ def create
+
+    # Instantiate server manager object and get matching site using resource name
+    inst_cmd = []
+    inst_cmd << '[Void][Reflection.Assembly]::LoadWithPartialName("Microsoft.Web.Administration");'
```
Would prefer this in an erb instead of a string array

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -38,6 +38,59 @@ def create
+    inst_cmd << '[Void][Reflection.Assembly]::LoadWithPartialName("Microsoft.Web.Administration");'
+    inst_cmd << '$serverManager = New-Object Microsoft.Web.Administration.ServerManager;'
+    inst_cmd << "$site = $serverManager.Sites | where { $_.Name -eq '#{@resource[:name]}' };"
+    inst_cmd << "$app = $site.Applications | where { $_.Path -eq '/' };"
```
These where clauses need to be tested on servers with more than a few sites. In prior testing with instances and create/delete, these methods get significantly slower as the number of sites go up

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -38,6 +38,59 @@ def create
+          inst_cmd << "$site.#{property.name.to_s}='#{property.value}';"
+      end
+    end
+    inst_cmd << '$serverManager.CommitChanges();'
```
This needs a try/catch around it so we can present an error message

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -47,8 +100,11 @@ def destroy
   def exists?
-    inst_cmd = "Get-Website -Name '\"#{@resource[:name]}\"'"
-    result   = self.class.run(inst_cmd)
+    Puppet.err('SITE Exists Called')
```
All of these err calls either need to be changed to debug or removed

jpogran commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -107,33 +163,36 @@ def self.prefetch(resources)
   def self.to_bool(value)
-    return true   if value == true   || value =~ (/(true|t|yes|y|1)$/i)
-    return false  if value == false  || value =~ (/(false|f|no|n|0)$/i)
+    return true if value == true || value =~ (/(true|t|yes|y|1)$/i)
```
Its generally bad form to change whitespace without change to code, harder to read diffs

hunner commented on lib/puppet/provider/iis_site/webadministration.rb:
```diff
@@ -47,8 +61,8 @@ def destroy
   def exists?
-    inst_cmd = "Get-Website -Name '\"#{@resource[:name]}\"'"
-    result   = self.class.run(inst_cmd)
+    inst_cmd = "Get-Website -Name #{@resource[:name]}"
```
This change breaks sites with spaces in the name

wilson208 replied:
Thanks @hunner, good catch!

hunner committed:
Fix calling update twice

Commit:
(maint) Small fixes and some cleanup of iis_site self.instances method
- Fix errors when no iis_site resources exist
- Fix typo causing self.instances to return empty physicalpath's
- In iis_site’s self.to_bool, handle case of empty value returns false in existing regex.
- In iis_sites’s self.to_bool, return true or false symbols

Commit:
(maint) Fix logproperties.ps1 template to correctly set array of flags
as a comma-separated string.

Overwrite logflags insync? method to deal with differences in ordering

Commit:
(MODULES-3887) Implement flush for iis_site resource’s web administration provider

By reusing a lot of the same powershell templates already used in creating a website. Added one new template, setwebsite.ps1, which contains the same Try-SetItemProperty as in newwebsite.ps1 and then reuses the generalproperties.ps1, logproperties.ps1 & serviceautostartprovider.ps1 templates.

Commit:
(maint) Fixes to acceptance test support files
- Allow puppet_resource_should_show matcher to handle matching on arrays
- Fix puppet_resource_should_show matcher throwing error when matching on a directory path by escaping value
- Fix ‘an idempotent resource’ example not raising error when changes are made during 2nd puppet run

Commit:
(MODULES-3887) Add acceptance tests for iis_site resource and utility file with helper methods

Added the following utility methods:
 - create_path(path) - Creates a directory using powershell if it doesn't exist
 - remove_all_sites - Removes all sites from iis
 - create_site(name, started, path) - Creates given path, creates a website with given physicalpath and name and sets that site to either started or stopped, based on started param

Add the following tests:
 - Create a site with basic params
 - Create a site with all params specified
 - - Using W3C log format, logflags and logtruncatesize
 - - Using NCSA log format and logperiod rather than logtruncatesize
 - Change site state
 - - From Stopped to Started
 - - From Started to Stopped
 - - From Started to Absent
 - Ensure run fails with invalid params for:
 - - logformat param
 - - logperiod param
 - Ensure can update existing, previously set param:
 - - physicalpath param
 - - enabledprotocols param
 - - logflags param

wilson208 commented:
@jpogran @glennsarti I have updated this PR with what we talked about last night about the performance of the powershell I was using. I have broken this down into smaller commits that detail everything I did. All spec & acceptance tests are now passing also.

jpogran commented on lib/puppet/provider/templates/webadministration/_setwebsite.ps1.erb:
```diff
@@ -0,0 +1,24 @@
+  ensure          = '<%= "#{resource[:ensure]}" %>'
+}
+
+function Try-SetItemProperty
```
I would rather see this function pulled out of _newwebsite.ps1.erb into a general erb rather than duplicate it here

wilson208 replied:
Have pulled this out now to trysetitemproperty.ps1.erb and pulling it in within the create and update methods.

jpogran commented on lib/puppet/type/iis_site.rb:
```diff
@@ -206,25 +206,26 @@
-        'BytesRecv','TimeTaken','ServerPort','UserAgent','Cookie','Referer',
-        'ProtocolVersion','Host','HttpSubStatus'
-        ].include?(value)
+          'Date','Time','ClientIP','UserName','SiteName','ComputerName','ServerIP',
```
I think you are adding tabs here? Only include needs to be moved

wilson208 replied:
So much for me 'cleaning up' my commits, missed this, sorted it now.

jpogran commented on spec/support/examples/idempotent_resource.rb:
```diff
@@ -1,9 +1,9 @@
 shared_examples 'an idempotent resource' do
   it 'should run without errors' do
-    apply_manifest(@manifest, :expect_changes => true)
+    apply_manifest(@manifest, :catch_failures => true)
```
This is used by other tests. What does this change do to those?

wilson208 replied:
This change was not necessary, however in a lot of our other modules tests we are using `:catch_changes` and `:catch_failures` rather than `:expect_changes` and `:expect_failures` and I was changing below from `:expect_changes => false` to `:catch_changes => true` and thought it best that I change this also for the sake of consistency.

wilson208 replied:
However this will not affect other tests.

jpogran commented on spec/support/examples/idempotent_resource.rb:
```diff
@@ -1,9 +1,9 @@
 
   it 'should run a second time without changes' do
-    apply_manifest(@manifest, :expect_changes => false)
+    apply_manifest(@manifest, :catch_changes => true)
```
This is used by other tests. What does this change do to those?

wilson208 replied:
This check for idempotency was broken previously and errors were not being raised, even if changes were made on the 2nd run of the given manifest.

jpogran commented on spec/support/matchers/puppet_resource_should_show.rb:
```diff
@@ -4,8 +4,10 @@ def puppet_resource_should_show(property_name, value)
+              /(#{property_name})(\s*)(=>)(\s*)(\[#{value.sort.map!{ |v| "'#{v}'" }.join(', ')}\])/i
             else
-              /(#{property_name})(\s*)(=>)(\s*)('#{value}'|#{value})/i
+              /(#{property_name})(\s*)(=>)(\s*)('#{Regexp.escape(value)}'|#{Regexp.escape(value)})/i
```
This is used by other tests. What does this change do to those?

wilson208 replied:
This will not affect other tests at current. The issue was that for the first time in our tests we are ensuring a path property is set correctly and hence we have backslashes in the path that need to be escaped otherwise the regex will fail. This Regexp.escape function is specifically for escaping strings so they can be used in regular expressions and as such should not cause any issues.

jpogran commented on spec/support/utilities/iis_site.rb:
```diff
@@ -0,0 +1,17 @@
+
+def remove_all_sites
+  on(default, "C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe -noninteractive -noprofile -executionpolicy bypass -command \"& {Get-Website | Remove-Website}\"")
+end
```
Nit: no newline at end of file

jpogran commented:
👍
